### ISSUE TITLE
How to set up vhosts to use fastcgi or php-fpm?

### ISSUE THREAD

**Issue by osopolar** [phpbrew/phpbrew#478]
May anybody help me with getting two php versions running on the same server (ubuntu 14.10; for development only). Ideally I would use the system default for one site and php 5.2.17 (provided by phpbrew) for the other site.

There is already a ticket under [#127 Running PHPs in parallel on single web server?](/phpbrew/phpbrew/issues/127), which says I need php-fpm. I finally compiled php 5.2.17 with php-fpm support, but I don't know what I have to do next. How to configure the fastcgi for the virtual host? I am not sure if it is necessary to use php-fpm and therefore call `phpbrew fpm start` each time I boot the server, or if just fastcgi is fine too.

Would be nice to have a small example of what I have to add to the vhost configuration to get it work.

**Comment by morozov** [comment]
I run many PHPs at a time on a single Apache server using vhosts. Here's my config example:
```
ServerName localhost

<Directory /var/www>
    Options Indexes FollowSymLinks
    AllowOverride All
    Require all granted
</Directory>

# This is configuration for localhost
<VirtualHost *:80>
    DocumentRoot /var/www
    <FilesMatch \.php$>
        SetHandler "proxy:unix:/home/morozov/.phpbrew/php/php-5.6.6/var/run/php-fpm.sock|fcgi://localhost"
    </FilesMatch>
</VirtualHost>

# Per-project PHP version
<VirtualHost *:80>
    ServerName a-project-on-53.local
    DocumentRoot /var/www/a-project-on-53/public
    <FilesMatch \.php$>
        SetHandler "proxy:unix:/home/morozov/.phpbrew/php/php-5.3.29/var/run/php-fpm.sock|fcgi://localhost"
    </FilesMatch>
</VirtualHost>

# Other projects
<VirtualHost *:80>
    ServerName local
    ServerAlias *.local
    VirtualDocumentRoot /var/www/%1/public
    <FilesMatch \.php$>
        SetHandler "proxy:unix:/home/morozov/.phpbrew/php/php-5.6.6/var/run/php-fpm.sock|fcgi://localhost"
    </FilesMatch>
</VirtualHost>
```

You need to do `phpbrew fpm start` each time you boot your server for every PHP version you use.

**Comment by osopolar** [comment]
Thanks for your example config. Due to the lack of response I figured it out on my own. Actually it was a loooong way to get it work (with and without php-fpm). Finally I use fastcgi, but not php-fpm. By default the virtual host uses the systems php and if I need a different php version I link to the phpbrew version. I do not net to start fpm after booting.

So my host file (/etc/apache2/sites-enabled/example.dev.conf) looks like:

```
<IfModule mod_fastcgi.c>
# see http://www.fastcgi.com/mod_fastcgi/docs/mod_fastcgi.html#fastcgiserver
FastCgiServer /var/www/php-fcgi/php52 -processes 1 -idle-timeout 30
</IfModule>

<VirtualHost *:80>
  ServerName example.dev
  ServerAlias *.example.dev
  DocumentRoot /home/dev/websites/example.dev
  <Directory /home/dev/websites/example.dev>
    Options Indexes FollowSymLinks MultiViews
    AllowOverride All
    # Apache < 2.4
    Order allow,deny
    allow from all
    # Apache 2.4
    Require all granted
  </Directory>

  <IfModule mod_fastcgi.c>
    AddType application/x-httpd-fastphp5 .php .html .htm
    Action application/x-httpd-fastphp5 /php5-fcgi-php-5.2.17
    Alias /php5-fcgi-php-5.2.17 /var/www/php-fcgi/php52
    <Directory /var/www/php-fcgi/php5>
      Require all granted
    </Directory>
  </IfModule>
</VirtualHost>
```

and the /etc/apache2/mods-enabled/fastcgi.conf looks like:

```
<IfModule mod_fastcgi.c>
  AddType application/x-httpd-fastphp5 .php
  Action application/x-httpd-fastphp5 /php5-fcgi-php-system
  Alias /php5-fcgi-php-system /var/www/php-fcgi/php5-system
  FastCgiServer /var/www/php-fcgi/php5-system -processes 1 -idle-timeout 310 -flush
  <Directory /var/www/php-fcgi/php5>
    Require all granted
  </Directory>
</IfModule>
```

And the file /var/www/php-fcgi/php52:

```
#!/bin/sh

PHPRC=/home/dev/.phpbrew/php/php-5.2.17/etc/cgi
export PHPRC
export PHP_FCGI_MAX_REQUESTS=5000
export PHP_FCGI_CHILDREN=8
exec /home/dev/.phpbrew/php/php-5.2.17/bin/php-cgi
```

The php-5.2.17 I builded this way:

```
phpbrew download --mirror http://museum.php.net/php5/ 5.2.17
cd ~/.phpbrew/build/php-5.2.17/
curl -s https://mail.gnome.org/archives/xml/2012-August/txtbgxGXAvz4N.txt | patch -p0
curl -s "https://bugs.php.net/patch-display.php?bug_id=54736&patch=debian_patches_disable_SSLv2_for_openssl_1_0_0.patch&revision=1305414559&download=1" | patch -p1
# http://php-fpm.org/
curl -s http://php-fpm.org/downloads/php-5.2.17-fpm-0.5.14.diff.gz | gunzip -c | patch -p1
CXX=g++ /usr/bin/phpbrew --debug install --old 5.2.17 +bcmath+bz2+cli+ctype+curl+dba+exif+fpm+gd+gettext+iconv+imap+intl+mcrypt+mhash+mysql=/usr/bin/mysql_config+openssl+pdo+mbstring+mbregex+soap+wddx+zlib -- --enable-force-cgi-redirect --enable-fastcgi
```

and php-5.3.29 (without +fpm as php-fpm and php-cgi are mutually exclusive and other tiny differences): `/usr/bin/phpbrew --debug install php-5.3 +default+cgi+intl+gd+mysql -- --enable-cgi`

**Comment by akomm** [comment]
Do you use fast-cgi because you got it is a solution proposal or because you need it?
Ubuntu 14.10 should have Apache 2.4+ installed. Since 2.4 you can use LoadModule directive in **virtual host** context:  
http://httpd.apache.org/docs/2.4/mod/mod_so.html#LoadModule

That means, you can add different LoadModule directives in your virtual hosts, pointing to different php versions.
Example: LoadModule php5_module /usr/lib/apache2/modules/libphp5.4.40.so

I did not test it, but by documentation it should work. I see there no constraints or exceptions.

**Comment by osopolar** [comment]
Ah thanks I thought that it was not possible, I'll test it.

**Comment by akomm** [comment]
You might have to disable the default mod entry (outside virtual host context). Not sure if the LoadModule directive overwrite previous definitions for the same module, before loading them or not.

**Comment by morozov** [comment]
> Since 2.4 you can use LoadModule directive in *virtual host* context

@akomm if I'm not mistaken, it's still not possible to load multiple PHP modules at a time (one for each version) at least because they have the same name. I tried patching them and naming them differently and recompiling but still didn't make it to load more than one module. The fast-cgi configuration just works.

Please share you config if you make it to use modules.

**Comment by akomm** [comment]
You'r right. You have to load them all at once and switch the usage per vhost. The problem is:
LoadModule [name] [DSO]
Name must be a symbol in the DSO. This is php5_module for all the versions. That is why you can not just rename it like that: php53_module. You have to compile the DSO so that each of them has a different symbol name. Looks like it is possible using the -n option (apxs):

http://httpd.apache.org/docs/2.4/programs/apxs.html
```
apxs -g [ -S name=value ] -n modname
```

I'm going to test it.

**Comment by c9s** [comment]
Hi all, 

there is a new command "fpm setup" integrated into phpbrew in 1.22 release.  the following daemon supports:

- `--systemctl`
- `--initd`
- `--launchctl`

I don't have a chance to try the second & the third option yet, if you're interested in, please give it a shot. :D

**Comment by dimul** [comment]
> phpbrew download --mirror http://museum.php.net/php5/ 5.2.17
> cd ~/.phpbrew/build/php-5.2.17/

How this execution is valid?

`phpbrew download` will just download a tarbal, how is it possible to cd into none existing build directory? Create it manually?

**Comment by morozov** [comment]
@dimul - I'm closing this ticket as the original request seems resolved. Please feel free to file another one regarding your problem.
